actions/promote/createBatch.js

```diff
@@ -0,0 +1,170 @@
+                batchesInfo: batchManager.getBatchesInfo()
+            });
+            logger.info(payload);
+
+            // Finalize and Trigger N Track the batches
+            await batchManager.finalizeInstance({
+                adminPageUri, projectExcelPath, fgRootFolder, doPublish
+            });
+            logger.info('Instance finalized and started');
+        }
+    } catch (err) {
+        await fgStatus.updateStatusToStateLib({
+            status: FgStatus.PROJECT_STATUS.COMPLETED_WITH_ERROR,
+            statusMessage: err.message,
+        });
+        logger.error(err);
+        payload = err;
+    }
+
+    return {
+        body: payload,
+    };
+}
+
+/**
+ * Find all files in the pink tree to promote.
+ */
+async function findAllFiles() {
+    const { sp } = await getConfig();
+    const baseURI = `${sp.api.excel.update.fgBaseURI}`;
+    const rootFolder = baseURI.split('/').pop();
+    const options = await getAuthorizedRequestOption({ method: 'GET' });
+    // Temporarily restricting the iteration for promote to under /drafts folder only
+    return findAllFloodgatedFiles(baseURI, options, rootFolder, [], ['/drafts']);
+}
+
+/**
+ * Iteratively finds all files under a specified root folder.
+ */
+async function findAllFloodgatedFiles(baseURI, options, rootFolder, fgFiles, fgFolders) {
+    while (fgFolders.length !== 0) {
+        const uri = `${baseURI}${fgFolders.shift()}:/children?$top=${MAX_CHILDREN}`;
+        // eslint-disable-next-line no-await-in-loop
+        const res = await fetchWithRetry(uri, options);
+        if (res.ok) {
+            // eslint-disable-next-line no-await-in-loop
+            const json = await res.json();
+            // eslint-disable-next-line no-await-in-loop
+            const driveItems = json.value;
+            driveItems?.forEach((item) => {
+                const itemPath = `${item.parentReference.path.replace(`/drive/root:/${rootFolder}`, '')}/${item.name}`;
+                if (item.folder) {
+                    // it is a folder
+                    fgFolders.push(itemPath);
+                } else {
+                    const downloadUrl = item['@microsoft.graph.downloadUrl'];
+                    fgFiles.push({ fileDownloadUrl: downloadUrl, filePath: itemPath });
+                }
+            });
+        }
+    }
+
+    return fgFiles;
+}
+
+/**
+ * Create batches based on configs and files to process
+ * @param {*} batchManager BatchManager for creating batches.
+ */
+async function createBatch(batchManager) {
+    let payload = 'Getting all floodgated files to promote.';
+    // Iterate the floodgate tree and get all files to promote
+    const allFgFiles = await findAllFiles();
+    logger.info(`Total files to process ${allFgFiles?.length}`);
+    // create batches to process the data
+    for (let i = 0; i < allFgFiles.length; i += 1) {
+        // eslint-disable-next-line no-await-in-loop
+        await batchManager.addFile(allFgFiles[i]);
```

Comment:
Instead of iterating of the `allFgFiles` array, perhaps we can slice the array into chunks of 150 or `maxFilesPerBatch` value?

Another option is to call `addFile` as part of the tree iteration itself. Basically replacing `fgFiles.push({ fileDownloadUrl: downloadUrl, filePath: itemPath });` at line 143.